### ISSUE TITLE
allow multiple OSC-MKtls from same netaddr

### ISSUE THREAD

**Issue by LFSaw** [ModalityTeam/Modality-toolkit#245]
This is especially needed for (communication) devices that are serving for multiple controllers/actuators.
A good example are the SenseStage miniBees:

```
// works since 5a99d3bd02fc5f300c9f0d822f4838d23420e3f0
q.bee1 = MKtl(\bee1, "*1*xpee*"); // prints initialisation message twice (?)
q.bee1.mktlDevice.updateSrcAddr("localhost", 52029); // set ip and port according to (arbitrarily) chosen sender port. this is not the reason for the issue.

// tracing
q.bee1.trace; 
q.bee1.trace(false);

// initiating MiniBee with ID 2 results in a massive error call stack
q.bee2 = MKtl(\bee2, "*2*xpee*");

// also rebuilding does result in the same call stack 
MKtl(\bee2).rebuild("*2*xpee*");
```

```
ERROR: Message 'at' not understood.
RECEIVER:
   nil
ARGS:
   Symbol 'srcPort'
PATH: miniBees-xpee.scd
CALL STACK:
	DoesNotUnderstandError:reportError   0x1174fed98
		arg this = <instance of DoesNotUnderstandError>
	Nil:handleError   0x117500de8
		arg this = nil
		arg error = <instance of DoesNotUnderstandError>
	Thread:handleError   0x116e1f968
		arg this = <instance of Thread>
		arg error = <instance of DoesNotUnderstandError>
	Object:throw   0x11753e408
		arg this = <instance of DoesNotUnderstandError>
	Object:doesNotUnderstand   0x11d21f7f8
		arg this = nil
		arg selector = 'at'
		arg args = [*1]
	< FunctionDef in Method Meta_MKtlLookup:findByIDInfo >   0x11d34f828
		arg key = 'srcPort'
		arg value = 12345
		var hereIDval = nil
	Dictionary:keysValuesArrayDo   0x118af3b38
		arg this = <instance of Event>
		arg argArray = [*12]
		arg function = <instance of Function>
		var i = 2
		var j = 1
		var key = nil
		var val = nil
		var arraySize = nil
	Dictionary:keysValuesDo   0x11753ac58
		arg this = <instance of Event>
		arg function = <instance of Function>
	< FunctionDef in Method Meta_MKtlLookup:findByIDInfo >   0x118af4698
		arg hereInfo = <instance of Event>
		var hereIdInfo = nil
		var hereIdDict = nil
		var matches = true
	< FunctionDef in Method Dictionary:select >   0x116e11478
		arg key = 'osc_bee1_16909060_12345'
		arg elem = <instance of Event>
	Dictionary:keysValuesArrayDo   0x118af51f8
		arg this = <instance of Event>
		arg argArray = [*8]
		arg function = <instance of Function>
		var i = 2
		var j = 1
		var key = nil
		var val = nil
		var arraySize = nil
	Dictionary:keysValuesDo   0x1175021d8
		arg this = <instance of Event>
		arg function = <instance of Function>
	Dictionary:select   0x11753ff88
		arg this = <instance of Event>
		arg function = <instance of Function>
		var res = <instance of Event>
	Meta_MKtlLookup:findByIDInfo   0x118aee578
		arg this = <instance of Meta_MKtlLookup>
		arg inIdInfo = <instance of Event>
		var matches = true
		var matching = nil
		var ordered = nil
		var inIdDict = <instance of Event>
	Meta_MKtlDevice:open   0x11910b4a8
		arg this = <instance of Meta_MKtlDevice>
		arg name = 'bee2'
		arg parentMKtl = <instance of MKtl>
		arg multiIndex = nil
		var lookupName = nil
		var lookupInfo = nil
		var protocol = 'osc'
		var idInfo = <instance of Event>
		var desc = <instance of MKtlDesc>
		var subClass = nil
		var newDevice = nil
		var deviceCandidates = nil
	MKtl:openDevice   0x117535648
		arg this = <instance of MKtl>
		arg lookAgain = false
		arg multiIndex = nil
		var protocol = 'osc'
...
^^ The preceding error dump is for ERROR: Message 'at' not understood.
RECEIVER: nil
```

**Comment by adcxyz** [comment]
General answer - 
That shows an issue that arises in OSC/MKtl
+ we do not know beforehand which devices are there 
(unless they are sending already and OSCMonitor finds them and identifies them, which does not happen ATM)
+ we do not know beforehand what the NetAddresses will be, so we cannot use 
them in the descfile idInfo entry to find the actual device(s). 
The way this is handled now is that you update the addresses afterwards; 
one could also pre-declare the oscdevices to MKtlLookup by doing 
```
MKtlLookup.addOSC(sendAddr, deviceName, replyAddr)
```
and see whether they are found properly then. 
which one should be recommended practice?

**Comment by adcxyz** [comment]
Specific answer:
what goes wrong here is that the existing first xbee is tested by MKtlLookup, 
and that fails because the entry there does not to have the info needed to identify it yet.  
this should have happened in MKtlLookup.addOSC. 
however, a problem with that is that keys are made for the (dummy) netaddresses 
given in the desc files, and these are tricky to change after the fact. 
That actually suggests doing ```MKtlLookup.addOSC(sendAddr, deviceName, replyAddr)``` first is better.
Or much preferably, rethink how idInfo works for OSC mktls - don't use netaddrs in the descs as they will change, think of robust alternatives.

// after quick fix in cceb7ae, this works just well enough so one can test working with multiple devices
```
q = q ? ();
q.bee1 = MKtl(\bee1, "*1*xpee*"); // prints initialisation message twice (?); 
// -> yes, should only post once.
q.bee1.mktlDevice.updateSrcAddr("localhost", 52029); 

// initiating MiniBee with ID 2 works now when multiIndex is given:
q.bee2 = MKtl(\bee2, "*2*xpee*", multiIndex: 1);
```

**Comment by LFSaw** [comment]
Thanks for the quick fix, will try later today.

> rethink how idInfo works for OSC mktls - don't use netaddrs in the descs as they will change, think of robust alternatives.

I think this is the preferred way... however it is very hard to come up with a robust alternative... there's nothing one can get from an OSC device. But this might mean that we can as well use the ```deviceName```, right?

**Comment by adcxyz** [comment]
Finding the proper desc (which is one of the issues here) is IMO best done by deviceName alone, so should be changed to that. Exception - sending devices from localhost (e.g. aManta), keep the same address always.

Identifying individual sending osc devices only works by netaddress data, otherwise they are indistinguishable, no? So that id strategy is OK until they accidentally get a new address; 
if that happens, we can still identify them by argTemplate (if they have hardware-side id numbers set); 
or we can track incoming traffic, identify by message names, find candidates for which old IP moved to which new one, and replace on the fly. this would then best live in an OSCTrafficRegulator utility.

**Comment by adcxyz** [comment]
// As an update to Tills example starting this issue, 
// here is a working example for making two minibees and distinguishing them:
(
~bee1 = MKtl(\bee1, "*minibee1");
~b1addr = NetAddr("192.168.100.105", 54615);
// update with the address as seen in OSCMonitor or OSCFunc.trace:
~bee1.mktlDevice.updateSrcAddr(~b1addr.hostname, ~b1addr.port);
)

// func for faking OSC message from a given NetAddr: 
~fakeSendFrom = { |netaddr, msg, recvPort| 
	recvPort = recvPort ?? NetAddr.langPort;
	thisProcess.recvOSCmessage(nil, netaddr, recvPort, msg);
};

OSCFunc.trace;
~fakeSendFrom.(~b1addr, ['/test', 1, 2, 3]);
OSCFunc.trace(false);

~bee1.trace;
~bee1.mktlDevice.oscFuncDictionary[\accel].argTemplate;

// this fake-sends from minibee1 instance 1 (argTemplate 1) 
~fakeSendFrom.(~b1addr, ['/minibee/data', 1, 1, 2, 3]);
// posts: MKtl('bee1') osc accel > [ 0.667, 0.833, 1 ] | type: nil

// different argTemplate value - no post
~fakeSendFrom.(~b1addr, ['/minibee/data', 2, 1, 2, 3]);
// wrong address, no post
~fakeSendFrom.(NetAddr.localAddr, ['/minibee/data', 1, 1, 2, 3]);

(
// make a second minibee1 - use multiIndex
~bee2 = MKtl(\bee2, "*minibee1", false, 1);
// give it a different NetAddr
~b2addr = NetAddr("192.168.100.111", 54615);
~bee2.mktlDevice.updateSrcAddr(~b2addr.hostname, ~b2addr.port);
)
~bee2.trace;

// distinguish by sending address (same argTemplate)
~fakeSendFrom.(~b2addr, ['/minibee/data', 1, 1, 2, 3]);
~fakeSendFrom.(~b1addr, ['/minibee/data', 1, 1, 2, 3]);

// try distinguishing by argTemplate as well : 
// - assuming that this is an ID for used for multiple devices
// put different value in the argTemplate: 
~bee2.mktlDevice.oscFuncDictionary[\accel].argTemplate.put(0, 2);
~fakeSendFrom.(~b2addr, ['/minibee/data', 2, 1, 2, 3]);
~fakeSendFrom.(~b2addr, ['/minibee/data', 1, 1, 2, 3]);
```

**Comment by LFSaw** [comment]
Thanks for this, @adcxyz . you got my example slightly wrong, though:

The two minibees are coming from _the same_ NetAddr/Port configuration and are only separable by their ID in the OSC message (```/minibee/data [id] [d1] ... [dN] ```). 
So this is why I built two descriptions, one to be made with 

```
q.bee1 = MKtl(\bee1, "*1*xpee*"); 
```

one with

```
q.bee2 = MKtl(\bee2, "*2*xpee*"); 
```

Does this change anything in your code?

**Comment by LFSaw** [comment]
to answer my question: yes it does, apparently:
This is now sufficient:
```
(
q.beeAddr = NetAddr("192.168.100.105", 59170);

q.bee1 = MKtl(\bee1, "*1*xpee*");
q.bee1.mktlDevice.updateSrcAddr(q.beeAddr.addr, q.beeAddr.port);
q.bee2 = MKtl(\bee2, "*2*xpee*", false);
q.bee2.mktlDevice.updateSrcAddr(q.beeAddr.addr, q.beeAddr.port);
)

q.bee1.trace();
q.bee2.trace();

q.bee1.trace(false);
q.bee2.trace(false);
```

**Comment by adcxyz** [comment]
if they come from the exact same NetAddr and use the exact same message names, you can still distinguish them if they have different hardware IDs that they send in the message; by *setting* the ID in argTemplate, as in the end of the example, you do exactly that.

**Comment by adcxyz** [comment]
so does 5d53c47 fix this bug for you?

**Comment by adcxyz** [comment]
fixed and tested with cdf146d
